### PR TITLE
Added Pipeline transformer capabilities with MXNet

### PR DESCRIPTION
**Pull request by dmmiller612**
This allows the use of spark ml, which is the preferred route for the future of spark.

Note, I have not been able to test this, because after following the directions on the mxnet page, I am still unable to run scalapkg. It is based off of something similar that I did for dl4j. So if someone could test it, that would be awesome. If I get it building, I will let someone know.

https://spark.apache.org/docs/latest/ml-classification-regression.html

Example same as before (mxnet lives under transformers for pipeline):
```
val conf = new SparkConf().setAppName("MXNet")
val sc = new SparkContext(conf)

val mxnet = new MXNet()
  .setBatchSize(128)
  .setLabelName("softmax_label")
  .setContext(Context.cpu()) // or GPU if you like
  .setDimension(Shape(784))
  .setNetwork(network) // e.g. MLP model
  .setNumEpoch(10)
  .setNumServer(2)
  .setFeaturesCol("features") //adding for dataframe/datasets
  .setLabelCol("label") //adding for dataframe/datasets
  .setNumWorker(4)
  // These jars are required by the KVStores at runtime.
  // They will be uploaded and distributed to each node automatically
  .setExecutorJars(cmdLine.jars)

val training = spark.read.format("libsvm").load("data/mllib/sample_libsvm_data.txt")

val model  = mxnet.fit(training)
model.predict(someVector)
```

More importantly, it opens up the ability to use spark's classification evaluators, and to join pipeline stages.

### CODE REVIEW

**Comment by piiswrong** [comment]
@javelinjs @Ldpe2G

**Comment by dmmiller612** [comment]
I'll get my other computer out to see if I can get this running, so it can be tested, if there are any found issues. Long term, I would like to process the data from dataset level, rather than using this as a wrapper for the existing spark mxnet.

**Comment by javelinjs** [comment]
will look into it today.

**Comment by dmmiller612** [comment]
Finally got it building. There are a couple of issues. Will fix it soon.

**Comment by dmmiller612** [comment]
built

File: scala-package/spark/src/main/scala/ml/dmlc/mxnet/spark/MXNet.scala [hunk 1]
```diff
@@ -196,6 +196,7 @@ class MXNet extends Serializable {
         dataName = params.dataName, labelName = params.labelName))
     }.cache()
 
+
```
Comment (javelinjs) [inline scala-package/spark/src/main/scala/ml/dmlc/mxnet/spark/MXNet.scala]: remove this change please.

File: scala-package/spark/src/main/scala/ml/dmlc/mxnet/spark/transformer/MXNet.scala [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+import org.apache.spark.ml.{PredictionModel, Predictor}
+import org.apache.spark.mllib.linalg.Vector
+import org.apache.spark.mllib.regression.LabeledPoint
+import org.apache.spark.sql.DataFrame
+import org.slf4j.{Logger, LoggerFactory}
+
+/**
+  * MXNet Training On Spark
+  *
+  * @author Derek Miller
```
Comment (javelinjs) [inline scala-package/spark/src/main/scala/ml/dmlc/mxnet/spark/transformer/MXNet.scala]: author tags are not allowed in ASF projects

File: scala-package/spark/src/main/scala/ml/dmlc/mxnet/spark/transformer/MXNet.scala [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  */
+class MXNet extends Predictor[Vector, MXNet, MXNetModelWrap] {
+
+  private val logger: Logger = LoggerFactory.getLogger(classOf[MXNet])
+  private val p: MXNetParams = new MXNetParams
+  private var _featuresCol: String = _
+  private var _labelCol: String = _
+
+  override val uid : String = UUID.randomUUID().toString
```
Comment (javelinjs) [inline scala-package/spark/src/main/scala/ml/dmlc/mxnet/spark/transformer/MXNet.scala]: you can omit the type here.

**Comment by dmmiller612** [comment]
Realized that argmax was not being used for classification. I made the switch.

**Comment by piiswrong** [comment]
Can one of the admins verify this patch?

File: scala-package/spark/src/main/scala/ml/dmlc/mxnet/spark/transformer/MXNet.scala [hunk 4]
```diff
@@ -159,7 +155,12 @@ class MXNetModelWrap(sc: SparkContext, mxNet: MXNetModel, uuid: String)
     val probArrays = mxNet.predict(features)
     val prob = probArrays(0)
     val arr = prob.get.toArray
-    arr.max
+    if (arr.length == 1) {
+      arr(0)
+    } else {
+      arr.indexOf(arr.max)
```
Comment (javelinjs) [inline scala-package/spark/src/main/scala/ml/dmlc/mxnet/spark/transformer/MXNet.scala]: just to make sure, for classification, `predict` returns the class index?

Comment (dmmiller612) [inline scala-package/spark/src/main/scala/ml/dmlc/mxnet/spark/transformer/MXNet.scala]: yep, as a double, which is the spark trait contract.

**Comment by dmmiller612** [comment]
@javelinjs Sorry, I was playing around with this again, and made one last change, due to java interop issues with the "copy" method. All should be good now.

**Comment by javelinjs** [comment]
Good to me. But the Travis build stucks, could you help to do a force merge? @piiswrong

**Pull request merged**